src/pcre2_printint.c

```diff
@@ -442,6 +443,148 @@ return (PCRE2_SPTR)next_char;
 
 
 
+/*************************************************
+*       Print character class                    *
+*************************************************/
+
+/* Prints a character class, which must be either an OP_CLASS, OP_NCLASS, or
+OP_XCLASS.
+
+Arguments:
+  f            file to write to
+  code         pointer in the compiled code
+  utf          TRUE if re is UTF (will be FALSE if UTF is not supported)
+  before       text to print before
+  after        text to print after
+
+Returns:       nothing
+*/
+
+static void
+print_class(FILE *f, PCRE2_SPTR code, BOOL utf, const char *before, const char *after)
+{
+BOOL printmap, invertmap;
+PCRE2_SPTR ccode;
+int i;
+
+fprintf(f, "%s[", before);
+
+/* Negative XCLASS has an inverted map whereas the original opcodes have
+already done the inversion. */
+invertmap = FALSE;
+if (*code == OP_XCLASS)
+  {
+  ccode = code + LINK_SIZE + 1;
+  printmap = (*ccode & XCL_MAP) != 0;
+  if ((*ccode & XCL_NOT) != 0)
+    {
+    invertmap = TRUE;
+    fprintf(f, "^");
+    }
+  ccode++;
+  }
+else  /* CLASS or NCLASS */
+  {
+  printmap = TRUE;
+  ccode = code + 1;
+  }
+
+/* Print a bit map */
+if (printmap)
+  {
+  uint8_t inverted_map[32];
+  const uint8_t *map = (const uint8_t *)ccode;
+  if (invertmap)
+    {
+    /* Using 255 ^ instead of ~ avoids clang sanitize warning. */
+    for (i = 0; i < 32; i++) inverted_map[i] = 255 ^ map[i];
+    map = inverted_map;
+    }
+  for (i = 0; i < 256; i++)
+    {
+    if ((map[i/8] & (1u << (i&7))) != 0)
+      {
+      int j;
+      for (j = i+1; j < 256; j++)
+        if ((map[j/8] & (1u << (j&7))) == 0) break;
+      if (i == '-' || i == ']') fprintf(f, "\\");
+      if (PRINTABLE(i)) fprintf(f, "%c", i);
+        else fprintf(f, "\\x%02x", i);
+      if (--j > i)
+        {
+        if (j != i + 1) fprintf(f, "-");
+        if (j == '-' || j == ']') fprintf(f, "\\");
+        if (PRINTABLE(j)) fprintf(f, "%c", j);
+          else fprintf(f, "\\x%02x", j);
+        }
+      i = j;
```

Comment:
It seems all code is moved here. Is this correct?

Reply:
Yes. I lifted out the existing code for printing character classes into a function.

Similarly, much of the "new" code in pcre2_compile_class.c is lifted out of pcre2_compile.c.